src/jit/alloc.cpp

```diff
@@ -478,64 +469,103 @@ static ArenaAllocator s_theAllocator;
+//    to the pool.
+void PooledAllocator::destroy()
+{
+    assert(isInitialized());
+    assert(this == &s_pooledAllocator);
+    assert(s_pooledAllocatorState == POOLED_ALLOCATOR_IN_USE || s_pooledAllocatorState == POOLED_ALLOCATOR_SHUTDOWN);
+    assert(m_firstPage != nullptr);
+
+    // Free all but the first allocated page
+    for (PageDescriptor* page = m_firstPage->m_next, *next; page != nullptr; page = next)
     {
-        // Not initialized yet
+        next = page->m_next;
+        freeHostMemory(page);
+    }
 
-        bool res = s_theAllocator.initialize(memoryManager, true);
-        if (!res)
-        {
-            // failed to initialize
-            InterlockedExchange(&s_isPooledAllocatorInUse, 0);            
-            return nullptr;
-        }
+    // Reset the relevant state to point back to the first byte of the first page
+    m_firstPage->m_next = nullptr;
+    m_lastPage = m_firstPage;
+    m_nextFreeByte = m_firstPage->m_contents;
+    m_lastFreeByte = (BYTE*)m_firstPage + m_firstPage->m_pageBytes;
+
+    assert(getTotalBytesAllocated() == s_defaultPageSize);
 
-        s_pooledAllocator = &s_theAllocator;
-        
-        assert(s_pooledAllocator->getTotalBytesAllocated() == s_defaultPageSize);
-        s_pooledAllocator->mark(s_pooledAllocatorMark);    
+    // If we've already been shut down, free the first page. Otherwise, return the allocator to the pool.
+    if (s_pooledAllocatorState == POOLED_ALLOCATOR_SHUTDOWN)
+    {
+        ArenaAllocator::destroy();
     }
     else
     {
-        if (s_pooledAllocator->m_memoryManager != memoryManager)
-        {
-            // already initialize with a different memory manager
-            InterlockedExchange(&s_isPooledAllocatorInUse, 0);            
-            return nullptr;
-        }
+        InterlockedExchange(&s_pooledAllocatorState, POOLED_ALLOCATOR_AVAILABLE);
     }
-
-    assert(s_pooledAllocator->getTotalBytesAllocated() == s_defaultPageSize);
-    return s_pooledAllocator;
 }
 
 //------------------------------------------------------------------------
-// ArenaAllocator::returnPooledAllocator:
-//    Returns the pooled allocator after the callee has finished using it.
+// ArenaAllocator::getPooledAllocator:
+//    Returns the pooled allocator if it is not already in use.
 //
 // Arguments:
-//    allocator - The pooled allocator instance. This must be an instance
-//                that was obtained by a previous call to
-//                `ArenaAllocator::getPooledAllocator`.
-void ArenaAllocator::returnPooledAllocator(ArenaAllocator* allocator)
+//    memoryManager: The `IEEMemoryManager` instance in use by the caller.
+//
+// Return Value:
+//    A pointer to the pooled allocator if it is available or `nullptr`
+//    if it is already in use.
+//
+// Notes:
+//    Calling `destroy` on the returned allocator will return it to the
+//    pool.
+ArenaAllocator* ArenaAllocator::getPooledAllocator(IEEMemoryManager* memoryManager)
```

Comment:
You have cleaned up the destroy path at the use site below. 

It might be nice to do the same for the create path by creating a new non-pooled allocator here if the pooled one is not available.